### DISCUSSION TITLE
Generic fields based on keyof class types

### DISCUSSION THREAD

**Discussion by knovu** [MichalLytek/type-graphql#1465] [Q&A] [answered]
@MichalLytek, not sure if this is something feasible, but does type-graphql allow the ability to generate let’s say enums or allowed string types based on properties of ObjectType class?

TypeOrm has an interesting way in their FindOptions object that deals with checking if properties exist within a class.  I’ll have to dig an example up later, but they can at runtime determine if a property is a valid key within an ENTITY type, so if you try to select, filter, or order by from a table with an invalid object key, they throw an error.

the only way I can think of is making an enum object for every filtered query.  

I could also see this might clash with ObjectType properties who are those of field resolvers or essentially just relational fields to other tables in an ORM entity.

E.g. Given an abstract filter, I want to be able to generate allowed values dynamically for ObjectTypes being queried.

```ts
@ObjectType()
class Person {
 @Field()
 name: string:
}

function DynamicFilter<TItem>(TClassType: ClassType<TItem>) {
  @InputType({ isAbstract: true })
  abstract class Filter {
     @Field() // Here I want to set a type that’s like a property of the TItem DTO
     sortBy: keyof TItem;
  }
}
```

@E.g. the TypeORM way of getting said properties off an entity:

```ts
const properties = Object.keys(
   getConnection().getRepository(User).metadata.propertiesMap,
)
```

**Comment by MichalLytek** [comment, marked as answer]
The names of the properties of the object type fields exists in runtime but are hardly hardcoded in the metadata storage.
They are not exposed and not accessible for users.

Instead of making fancy solutions, just make an enum that matches properties of the object type and pass that as param to `DynamicFilter` - TS will do the rest of the checking of accessing those properties by name from enum.

**Reply by knovu** [reply]
After some thought into the design, yes you are correct, that was the best solution we came up with as well.

Having a custom decorator that basically would derive the fields, would cause some security issues, because what if we don't want certain fields to be exposed in the derived enum to whitelist strings that can be used.

**Reply by knovu** [reply]
I hope that makes sense, but basically, let's say you have a `DynamicFilter`, you don't want to allow sorting on a TypeORM DTO's properties that are just used for "relations" or something, or basic ID's, or anything internal that you dont want to be exposed API side.
